flex/engines/graph_db/grin/src/property/property.cc

```diff
@@ -0,0 +1,414 @@
+const char* grin_get_vertex_property_name(GRIN_GRAPH g, GRIN_VERTEX_TYPE vt,
+                                          GRIN_VERTEX_PROPERTY vp) {
+  auto _g = static_cast<GRIN_GRAPH_T*>(g);
+  auto& table = _g->get_vertex_table(vt);
+
+  const auto& name = table.column_name(vp & (0xff));
+  auto len = name.length() + 1;
+  char* out = new char[len];
+  snprintf(out, len, "%s", name.c_str());
+  return out;
+}
+
+GRIN_VERTEX_PROPERTY grin_get_vertex_property_by_name(GRIN_GRAPH g,
+                                                      GRIN_VERTEX_TYPE vt,
+                                                      const char* name) {
+  auto _g = static_cast<GRIN_GRAPH_T*>(g);
+  auto& table = _g->get_vertex_table(vt);
+  auto col = table.get_column(name);
+  if (col == nullptr) {
+    return GRIN_NULL_VERTEX_PROPERTY;
+  }
+  GRIN_VERTEX_PROPERTY vp;
+  vp = table.get_column_id_by_name(name);
+  vp += (vt * 1u) << 8;
+  vp += (_get_data_type(col->type()) * 1u) << 16;
+  return vp;
+}
+
+GRIN_VERTEX_PROPERTY_LIST grin_get_vertex_properties_by_name(GRIN_GRAPH g,
+                                                             const char* name) {
+  auto _g = static_cast<GRIN_GRAPH_T*>(g);
+  std::string prop_name(name);
+  auto vps = new GRIN_VERTEX_PROPERTY_LIST_T();
+  std::string _name = std::string(name);
+  for (auto idx = 0; idx < _g->schema().vertex_label_num(); idx++) {
+    auto& table = _g->get_vertex_table(static_cast<GRIN_VERTEX_TYPE>(idx));
+
+    auto col = table.get_column(name);
+
+    if (col != nullptr) {
+      GRIN_VERTEX_PROPERTY vp;
+      vp = table.get_column_id_by_name(name);
+      vp += (idx * 1u) << 8;
+      vp += (_get_data_type(col->type()) * 1u) << 16;
+      vps->emplace_back(vp);
+    }
+  }
+  if (vps->size() == 0) {
+    delete vps;
+    return GRIN_NULL_VERTEX_PROPERTY_LIST;
+  }
+  return vps;
+}
+#endif
+
+#ifdef GRIN_WITH_VERTEX_PROPERTY
+bool grin_equal_vertex_property(GRIN_GRAPH g, GRIN_VERTEX_PROPERTY vp1,
+                                GRIN_VERTEX_PROPERTY vp2) {
+  return vp1 == vp2;
+}
+
+void grin_destroy_vertex_property(GRIN_GRAPH g, GRIN_VERTEX_PROPERTY vp) {}
+
+/**
+ * @TODO add type for GRIN_VERTEX_PROPERTY_T
+ */
+GRIN_DATATYPE grin_get_vertex_property_datatype(GRIN_GRAPH g,
+                                                GRIN_VERTEX_PROPERTY vp) {
+  return (GRIN_DATATYPE) (vp >> 16);
+}
+
+int grin_get_vertex_property_value_of_int32(GRIN_GRAPH g, GRIN_VERTEX v,
+                                            GRIN_VERTEX_PROPERTY vp) {
+  auto _g = static_cast<GRIN_GRAPH_T*>(g);
+  auto label = v >> 32;
+  auto vid = v & (0xffffffff);
+  auto plabel = (vp >> 8) & (0xff);
+  auto pdt = (vp >> 16);
+  auto pid = vp & (0xff);
+  if (label != plabel || pdt != GRIN_DATATYPE::Int32) {
+    grin_error_code = INVALID_VALUE;
+    return 0;
+  }
+  auto& table = _g->get_vertex_table(plabel);
+  auto col =
+      std::dynamic_pointer_cast<gs::IntColumn>(table.get_column_by_id(pid));
```

Comment:
Try to avoid dynamic_pointer_cast in get_value apis, based on the performance benchmark of grin-v6d, this can be 20-30 times slower. Instead, you can cache these columns as void* and then static_cast them back to the type you need in these apis.